pkg/volume/csi/csi_mounter.go

```diff
@@ -377,30 +383,49 @@ func (c *csiMountMgr) TearDownAt(dir string) error {
 // 1) if fstype is "", then skip fsgroup (could be indication of non-block filesystem)
 // 2) if fstype is provided and pv.AccessMode == ReadWriteOnly and !c.spec.ReadOnly then apply fsgroup
 func (c *csiMountMgr) applyFSGroup(fsType string, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy) error {
-	if fsGroup != nil {
-		if fsType == "" {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, fsType not provided"))
-			return nil
-		}
+	featureGateEnabled := utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeSupportsFSGroup)
+
+	// If the feature gate is enabled and we're set to Always, proceed.
+	// Otherwise, check to ensure a fsGroup is defined to proceed
+	if (featureGateEnabled && c.SupportsFSGroup == storage.AlwaysSupportsFSGroup) ||
+		fsGroup != nil {
+
+		// If the feature gate isn't enabled, then we should continue with the
+		// existing behavior. If it is enabled, then check to see if we should Evaluate.
+		if !featureGateEnabled || c.SupportsFSGroup == storage.EvaluateSupportsFSGroup {
+			if fsType == "" {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, fsType not provided"))
+				return nil
+			}
 
-		accessModes := c.spec.PersistentVolume.Spec.AccessModes
-		if c.spec.PersistentVolume.Spec.AccessModes == nil {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, access modes not provided"))
-			return nil
-		}
-		if !hasReadWriteOnce(accessModes) {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, only support ReadWriteOnce access mode"))
-			return nil
-		}
+			accessModes := c.spec.PersistentVolume.Spec.AccessModes
+			if c.spec.PersistentVolume.Spec.AccessModes == nil {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, access modes not provided"))
+				return nil
+			}
+			if !hasReadWriteOnce(accessModes) {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, only support ReadWriteOnce access mode"))
+				return nil
+			}
+
+			if c.readOnly {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, volume is readOnly"))
+				return nil
+			}
 
-		if c.readOnly {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, volume is readOnly"))
-			return nil
+			err := volume.SetVolumeOwnership(c, fsGroup, fsGroupChangePolicy)
+			if err != nil {
+				return err
+			}
 		}
 
-		err := volume.SetVolumeOwnership(c, fsGroup, fsGroupChangePolicy)
-		if err != nil {
-			return err
+		// If the feature gate is enabled and we're set to Always apply,
+		// proceed without performing any of the typical checks.
+		if featureGateEnabled && c.SupportsFSGroup == storage.AlwaysSupportsFSGroup {
+			err := volume.SetVolumeOwnership(c, fsGroup, fsGroupChangePolicy)
```

Comment:
Prefer not to have multiple call points. Can we have both Always and Evaluate cases flow to the same call?